### PR TITLE
provider/nomad: Nomad provider for managing jobs

### PR DESCRIPTION
**Pull request by mitchellh**
This adds a new provider `nomad` with a single resource `nomad_job` that can be used for managing jobs within a [Nomad](https://www.nomadproject.io) cluster. 

The primary purpose for this provider is to manage system tasks or always-running service tasks that are used as part of cluster bootstrapping. For example, monitoring systems, core services, etc. Other jobs are usually submitted directly to Nomad (such as basic app deployment).

This is documented and tested (unit + acceptance). I rebased the commits so you can look at the `provider/nomad` commit to just see the provider separate from the vendoring.

Finally, this provider is built to work with Nomad 0.4.1. It likely won't work for earlier versions.

### CODE REVIEW

File: builtin/providers/nomad/provider_test.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	"testing"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/terraform"
+)
+
+// How to run the acceptance tests for this provider:
+//
+// - Obtain an official Nomad release from https://nomadproject.io
+//   and extra "nomad" somewhere
```
Comment (jamtur01) [inline builtin/providers/nomad/provider_test.go]: extract? `extract the nomad binary` perhaps?

File: builtin/providers/nomad/resource_job.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+	}
+	if prevId != "" && prevId != jobspecStruct.ID {
+		log.Printf(
+			"[INFO] Deregistering %q before registering %q",
+			prevId, jobspecStruct.ID)
+
+		log.Printf("[DEBUG] Deregistering job: %q", prevId)
+		_, _, err := client.Jobs().Deregister(prevId, nil)
+		if err != nil {
+			return fmt.Errorf("error deregistering previous job %q: %s", prevId, err)
```
Comment (jamtur01) [inline builtin/providers/nomad/resource_job.go]: previous feels odd here. `error deregistering job` would work fine?

File: website/source/docs/providers/nomad/index.html.markdown [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+resource "nomad_job" "monitoring" {
+    jobspec = "${file("${path.module}/jobspec.hcl")}"
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `address` - (Optional) The HTTP(S) API address of the agent to use. Defaults to "127.0.0.1:4646". `NOMAD_ADDR` can also be used.
```
Comment (jamtur01) [inline website/source/docs/providers/nomad/index.html.markdown]: `The HTTP(S) API address of the Nomad agent to use` - make it explicit to make it consistent. 

I'd probably also: `127.0.0.1:4646`.

Comment (jamtur01) [inline website/source/docs/providers/nomad/index.html.markdown]: `The "NOMAD_ADDR" environment can also be used.`? WIth `NOMAD_ADDR`.

File: builtin/providers/nomad/resource_job.go (hunk 2 above)
Comment (mitchellh) [inline builtin/providers/nomad/resource_job.go]: I don't want the user to be confused why we may be deregistering a job "foo" when the jobspec perhaps has job "bar". I've updated the wording to be a bit more explicit about that, pushing soon.

File: website/source/docs/providers/nomad/index.html.markdown (hunk 3 above)
Comment (jamtur01) [inline website/source/docs/providers/nomad/index.html.markdown]: Ditto with the `NOMAD_REGION` line.

File: website/source/docs/providers/nomad/r/job.html.markdown [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+description: |-
+  Manages a job registered in Nomad.
+---
+
+# nomad\_job
+
+Manages a job registered in Nomad.
+
+This can be used to initialize your cluster with system jobs, common services,
+and more. In day to day Nomad use it is common for developers to submit
```
Comment (jamtur01) [inline website/source/docs/providers/nomad/r/job.html.markdown]: The second sentence doesn't make sense to me? I think I'm either missing a Nomad concept or it's a little convoluted?

File: website/source/docs/providers/nomad/r/job.html.markdown [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+jobs to Nomad directly, but jobs that must always be running or system-level
+jobs can be registered with Terraform.
+
+## Example Usage
+
+```
+resource "nomad_job" "app" {
+    jobspec = "${file("${path.module}/job.hcl")}"
```
Comment (jamtur01) [inline website/source/docs/providers/nomad/r/job.html.markdown]: I doubt anyone is ever going to write a whole jobspec inline but maybe an example of a jobspec somewhere so folks know that it is they are looking for?

File: builtin/providers/nomad/provider_test.go [hunk 6]
```diff
@@ -11,7 +11,7 @@ import (
 // How to run the acceptance tests for this provider:
 //
 // - Obtain an official Nomad release from https://nomadproject.io
-//   and extra "nomad" somewhere
+//   and extra the "nomad" binary
```
Comment (jamtur01) [inline builtin/providers/nomad/provider_test.go]: Still missing the `t` :)

**Comment by mitchellh** [comment]
@jamtur01 Thanks! Addressed feedback (and fixed that typo), while adding more examples.

**Comment by jamtur01** [comment]
@mitchellh Very neat BTW! Does << EOT blah EOT work in HCL? I had no idea. I wonder if that's documented anywhere? I'll go look. :)

**Comment by mitchellh** [comment]
@jamtur01 Yes it does! You can even see it in "action" in the acceptance tests for this provider. It may not be documented actually... but I can do that separate from this PR.

**Comment by jamtur01** [comment]
@mitchellh It's actually in the syntax docs already. Had just missed it! Thanks.

**Comment by jen20** [comment]
@jamtur01 Yes - so does `<<-` to remove hanging indent (à la Ruby)

**Comment by apparentlymart** [comment]
I was looking at #8769 primarily so that Terraform could interface with Nomad in a way that doesn't defeat Nomad's own diffing mechanism.

Certainly not meaning to suggest that this has to be blocked by #8769, but just wanted to point it out in case it helps to think about 1) whether such behavior is desirable anyway, and 2) assuming yes, whether adding it later would require a breaking change here that might make it worth tweaking the design now. (I think probably not, but I'm not 100% sure...)

**Comment by mitchellh** [comment]
@apparentlymart I think in Nomad's cause that would be nice but doesn't block it since Nomad is idempotent. If you make a stylistic change that causes a plan to update the job, Nomad won't actually do anything once the job is resubmitted. I think that makes this okay.

**Comment by jamtur01** [comment]
@jen20 Do you mean `<<~` (that removes hanging indent in Ruby - `<<-` preserves whitespace.

**Pull request merged**